internal/driver/ipam.go

```diff
@@ -141,137 +142,154 @@ func (i IpamDriver) ReleasePool(request *ipam.ReleasePoolRequest) error {
-		if err != nil {
-			log.Errorln(err)
-			return nil, err
-		} else if reserved {
-			// succeed to aquire reserved ip
-			IPs = []caliconet.IP{{IP: net.ParseIP(request.Address)}}
-		} else {
-			// Docker allows the users to specify any address.
-			// We'll return an error if the address isn't in a Calico pool, but we don't care which pool it's in
-			// (i.e. it doesn't need to match the subnet from the docker network).
-			log.Debugln("Reserving a specific address in Calico pools")
-			ip := net.ParseIP(request.Address)
-			ipArgs := calicoipam.AssignIPArgs{
-				IP:       caliconet.IP{IP: ip},
-				Hostname: hostname,
-			}
 
-			err = i.client.IPAM().AssignIP(context.Background(), ipArgs)
-			if err != nil {
-				err = errors.Wrapf(err, "IP assignment error, data: %+v", ipArgs)
-				log.Errorln(err)
-				return nil, err
-			}
-			IPs = []caliconet.IP{{IP: ip}}
+		version = ipNet.Version()
+		if version == 4 {
+			poolV4 = []caliconet.IPNet{caliconet.IPNet{IPNet: ipNet.IPNet}}
+			numIPv4 = 1
+			log.Debugln("Using specific pool ", poolV4)
+		} else if version == 6 {
+			poolV6 = []caliconet.IPNet{caliconet.IPNet{IPNet: ipNet.IPNet}}
+			numIPv6 = 1
+			log.Debugln("Using specific pool ", poolV6)
 		}
 	}
 
+	// Auto assign an IP address.
+	// IPv4/v6 pool will be nil if the docker network doesn't have a subnet associated with.
+	// Otherwise, it will be set to the Calico pool to assign from.
+	var IPsV4 []caliconet.IP
+	var IPsV6 []caliconet.IP
+	if IPsV4, IPsV6, err = i.client.IPAM().AutoAssign(
+		context.Background(),
+		calicoipam.AutoAssignArgs{
+			Num4:      numIPv4,
+			Num6:      numIPv6,
+			Hostname:  hostname,
+			IPv4Pools: poolV4,
+			IPv6Pools: poolV6,
+		},
+	); err != nil {
+		log.Errorln("IP assignment error")
+		return caliconet.IP{}, err
+	}
+	IPs = append(IPsV4, IPsV6...)
+
 	// We should only have one IP address assigned at this point.
 	if len(IPs) != 1 {
-		err := errors.New(fmt.Sprintf("Unexpected number of assigned IP addresses. "+
-			"A single address should be assigned. Got %v", IPs))
-		log.Errorln(err)
-		return nil, err
+		return caliconet.IP{}, errors.Errorf("Unexpected number of assigned IP addresses. "+
+			"A single address should be assigned. Got %v", IPs)
 	}
+	return IPs[0], nil
+}
 
-	resp := &ipam.RequestAddressResponse{
-		// Return the IP as a CIDR.
-		Address: formatIPAddress(IPs[0]),
+func (i IpamDriver) requestSpecifiedIP(request *ipam.RequestAddressRequest) (caliconet.IP, error) {
+	var err error
+
+	// specified address requested, so will try assign from reserved pool, then calico pool
+	log.Println("Assigning specified IP from reserved pool first, then calico pools")
+
+	var hostname string
+	if hostname, err = osutils.GetHostname(); err != nil {
+		return caliconet.IP{}, err
 	}
 
-	logutils.JSONMessage("RequestAddress response", resp)
+	// try to aquire ip from reserved ip pool
+	var aquired bool
+	if aquired, err = i.ripam.AquireIPIfReserved(request.Address); err != nil {
```

Comment:
it may seem the function name doesn't need ```IPIfReserved``` postfix due to the ```ReservedIPManager``` name.

Reply:
ok with the "IP" part, but I think "IfReserved" is still need, 'cause by "Aquire" named only, caller may expecting encouter an error when the ip is not in reserved pool.